studies/flatten.py

```diff
@@ -0,0 +1,1008 @@
+
+# RawArray
+def RawArray_flatten(self, axis=0):
+    if axis < 0:
+        raise NotImplementedError
+    else:
+        raise IndexError("axis > list depth of structure")
+
+RawArray.flatten = RawArray_flatten
+
+def count_strides(array):
+    strides = []
+    def recurse(array):
+        if isinstance(array, Content):
+            strides.append(array.__len__())
+            return count_strides(array.__getitem__(0))
+        else:
+            return array
+
+    recurse(array)
+
+    return strides
+
+
+def compact_array(array, depth=-1):
+    data_items = []
+
+    def recurse(array, depth):
+        if isinstance(array, Content) and array.__len__() > 0:
+            if depth != 0:
+                for it in range(array.__len__()):
+                    recurse(array.__getitem__(it), depth - 1)
+        else:
+            data_items.append(array)
+
+    recurse(array, depth)
+
+    return data_items
+
+def index_array(array, depth=-1):
+    indices = []
+    offset = array.offset
+
+    def recurse(array, index, depth):
+        if isinstance(array, Content) and array.__len__() > 0:
+            if depth != 0:
+                for it in range(array.__len__()):
+                    index = array.offset + it
+                    recurse(array.__getitem__(it), index, depth - 1)
+        else:
+            indices.append(index)
+
+    recurse(array, offset, depth)
+
+    return indices
+
+def partition(tuple, begin, end):
+    pivot_idx = begin
+    for i in range(begin + 1, end + 1):
+        if tuple[i][0] <= tuple[begin][0]:
+            pivot_idx += 1
+            tuple[i], tuple[pivot_idx] = tuple[pivot_idx], tuple[i]
+    tuple[pivot_idx], tuple[begin] = tuple[begin], tuple[pivot_idx]
+    return pivot_idx
+
+def quick_sort_recursion(tuple, begin, end):
+    if begin >= end:
+        return
+    pivot_idx = partition(tuple, begin, end)
+    quick_sort_recursion(tuple, begin, pivot_idx - 1)
+    quick_sort_recursion(tuple, pivot_idx + 1, end)
+
+def quick_sort(tuple, begin = 0, end = None):
+    if end is None:
+        end = len(tuple) - 1
+
+    return quick_sort_recursion(tuple, begin, end)
+
+# Returns sorted indices
+def NumpyArray_argsort(self, axis=0):
+    if axis < 0:
+        raise NotImplementedError
+    else:
+        tuple = []
+        indices = index_array(self, len(self.shape))
+
+        for i in range(len(indices)): {
+            tuple.append((self.ptr[indices[i]], indices[i]))
+        }
+
+        sorted_tuple = quick_sort(tuple)
+        sorted_indices = []
+        for it in range(len(tuple)):
+            sorted_indices.append(tuple[it][1])
+
+        return NumpyArray.onedim(sorted_indices)
+
+NumpyArray.argsort = NumpyArray_argsort
+
+# NumpyArray
+def NumpyArray_flatten(self, axis=0):
+    if axis < 0:
+        raise NotImplementedError
+    else:
+        if self.shape[0] == 0:
+            return NumpyArray.onedim([])
+
+        compact_ptr = compact_array(array, len(self.shape))
```

Comment:
No, the structure of NumpyArray is one-to-one with the structure of a NumPy array. NumPy really does use two equal-length lists of integers for `shape` and `strides` (and this is one of the genius things about it). One way to see that is that this is the information contained in [pybind11's `buffer_info`](https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html):

```c++
struct buffer_info {
    void *ptr;
    ssize_t itemsize;
    std::string format;
    ssize_t ndim;
    std::vector<ssize_t> shape;
    std::vector<ssize_t> strides;    // same size as shape
};
```

(This is the reason that NumpyArray uses `ssize_t` for these vectors, although that's turned out to be quite a headache when everything else is `int64_t`.) pybind11 doesn't have an `offset`, but that's because I'm reference-counting the `ptr`, so i don't want a new value of `ptr` whenever we slice off the beginning of an array. I want the sliced version to share data with the unsliced version, and to link the reference counts I use the same `std::shared_ptr<void>`.

Here's what they mean:

   * `byteoffset`: the number of bytes past the base `ptr` where we _really_ start. This could have been measured in bytes or in itemsize, I chose bytes because we rarely need to interpret the data in a NumpyArray, just slice it and pass it along.
   * `shape[0]` is the number of rows.
   * `shape[1]` is the number of columns, _not_ `strides[0]`!
   * `shape[2]` is the number of items in the third dimension, etc.
   * `strides[0]` is the number of bytes to skip from the beginning of one row to the next.
   * `strides[1]` is the number of bytes to skip from the beginning of one column to the next.
   * `strides[2]` is the number of bytes to skip from the beginning of one third dimension element to the next, etc.

Thus, `shape.size() == strides.size()`.

For a compact (completely contiguous in memory, C-order) NumPy array, the `strides` is determined by `shape` and `itemsize`. For example, for a three-dimensional array,

```c++
strides[2] = itemsize;
strides[1] = strides[2] * shape[2];
strides[0] = strides[1] * shape[1];
```

Things get more interesting if we're looking at sliced views of an array. WIth the above structure, NumPy can avoid having to copy an array when applying any `start:stop:step` slice.

First, consider a compact array:

```python
>>> import numpy as np
>>> array = np.arange(2*3*5).reshape(2, 3, 5)
>>> array
array([[[ 0,  1,  2,  3,  4],
        [ 5,  6,  7,  8,  9],
        [10, 11, 12, 13, 14]],

       [[15, 16, 17, 18, 19],
        [20, 21, 22, 23, 24],
        [25, 26, 27, 28, 29]]])
>>> array.shape, array.strides
((2, 3, 5), (120, 40, 8))
```

Now when we slice the third dimension, we can skip an arbitrary number of inner elements while still viewing the same data:

```python
>>> array[:, :, :3]
array([[[ 0,  1,  2],
        [ 5,  6,  7],
        [10, 11, 12]],

       [[15, 16, 17],
        [20, 21, 22],
        [25, 26, 27]]])
>>> array[:, :, :3].shape, array[:, :, :3].strides   # note the different *shape*
((2, 3, 3), (120, 40, 8))
```

This has the same strides (same number of bytes to skip before the beginning of each inner array) but different shape.

Even reversing the direction of the inner arrays can be a view of the same data, by making one of the strides negative:

```python
>>> array[:, :, ::-1]
array([[[ 4,  3,  2,  1,  0],
        [ 9,  8,  7,  6,  5],
        [14, 13, 12, 11, 10]],

       [[19, 18, 17, 16, 15],
        [24, 23, 22, 21, 20],
        [29, 28, 27, 26, 25]]])
>>> array[:, :, ::-1].shape, array[:, :, ::-1].strides
((2, 3, 5), (120, 40, -8))
```

One particularly clever thing this can do is blow up a single number into a huge array. The following array consists of a single element in memory, viewed as an array (usually used to "broadcast" a number into an array):

```python
>>> array = np.lib.stride_tricks.as_strided(123, (2, 3, 5), (0, 0, 0))
>>> array
array([[[123, 123, 123, 123, 123],
        [123, 123, 123, 123, 123],
        [123, 123, 123, 123, 123]],

       [[123, 123, 123, 123, 123],
        [123, 123, 123, 123, 123],
        [123, 123, 123, 123, 123]]])
>>> array.shape, array.strides
((2, 3, 5), (0, 0, 0))
```

Really, it's all one number. If I assign at any position, they all change:

```python
>>> array[1, 2, 4] = 999
>>> array
array([[[999, 999, 999, 999, 999],
        [999, 999, 999, 999, 999],
        [999, 999, 999, 999, 999]],

       [[999, 999, 999, 999, 999],
        [999, 999, 999, 999, 999],
        [999, 999, 999, 999, 999]]])
```

Another nifty thing this can do is unify "C-order" with "Fortran-order". The following array is Fortran-ordered in memory, but behaves like any other NumPy array because it's the `shape` and `strides` that count:

```python
>>> array = np.array(np.arange(2*3*5).reshape(2, 3, 5), order="F")
>>> array
array([[[ 0,  1,  2,  3,  4],
        [ 5,  6,  7,  8,  9],
        [10, 11, 12, 13, 14]],

       [[15, 16, 17, 18, 19],
        [20, 21, 22, 23, 24],
        [25, 26, 27, 28, 29]]])
>>> array.shape, array.strides   # note the *decreasing* strides
((2, 3, 5), (8, 16, 48))
```

Algorithms that need to iterate over a NumPy array (for the most part, ours _do not_) can then walk in byte-steps of `stride[dim]` and not need to know whether the array is C-order, Fortran-order, or sliced in a weird way. That's the brilliant thing about NumPy; congrats to whoever thought of it.

As cool as that is, it relies heavily on the assumption that data are rectilinear—we can't use it in Awkward Array. NumPy arrays are passed in without changing their structure because maybe we won't need to alter it, and we're optimistically avoiding array-copies. When we do have to deal with it, we have three choices (from hardest to easiest):

   1. Figure out how to apply what we're trying to do to the whole `shape`-`strides` structure. This is what `NumpyArray::getitem_bystrides` does, which is for non-advanced indexing (single-integer `at` extractions, `start:stop:step` range-slices, or tuples of them).
   2. Call `NumpyArray::contiguous()` to not-copy if the array is already contiguous or make a contiguous copy, then apply what we're trying to do on contiguous, C-order data.
   3. Call `NumpyArray::toRegularArray()` to make the NumPy array contiguous and one-dimensional, stuffing the dimensional structure in nested RegularArrays.

Choices (2) and (3) involve the same number of copies (zero if already contiguous, one otherwise), but (3) makes the data easier to apply recursive algorithms to. I should have implemented RegularArray much sooner; it would have saved effort. In fact, I might _remove_ complexity from NumpyArray's implementation by replacing some situations where I chose (2) to the equivalent with (3).

Choice (1) is more efficient because it involves zero copies, but it's questionable whether the complexity is worth the speed. Since Awkward Arrays are declared to be immutable, we can at any time switch one choice for another. Doing so changes in-place semantics, but we've already declared that in-place semantics are an internal detail that users should not rely on (and `ak.Array` makes it difficult to do so).